### PR TITLE
Implement SD3 loss weighting

### PR DESCRIPTION
**Pull request by Slickytail**
The loss-weighting schemes for SD3 training were not implemented correctly, causing all of them to be non-functional. I went ahead and implemented the lognorm and cosmap schemes, just by using the density at those timesteps. Potentially, a better approach would be to sample the timestep according to that density in the first place.

The Mode scheme is much harder to implement -- there's a reason that they didn't include an explicit form for the density in the paper (I couldn't find one...), so I put in an error message if you try to use it for now. 

@sayakpaul @kashif

### CODE REVIEW

**Comment by kashif** [comment]
thanks @Slickytail checking

**Comment by bghira** [comment]
previous code

image


new code

image

File: examples/dreambooth/train_dreambooth_lora_sd3.py [hunk 1]
```diff
@@ -1483,16 +1495,15 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
...
-                elif args.weighting_scheme == "logit_normal":
-                    # See 3.1 in the SD3 paper ($rf/lognorm(0.00,1.00)$).
-                    u = torch.normal(mean=args.logit_mean, std=args.logit_std, size=(bsz,), device=accelerator.device)
-                    weighting = torch.nn.functional.sigmoid(u)
-                elif args.weighting_scheme == "mode":
-                    # See sec 3.1 in the SD3 paper (20).
-                    u = torch.rand(size=(bsz,), device=accelerator.device)
-                    weighting = 1 - u - args.mode_scale * (torch.cos(math.pi * u / 2) ** 2 - 1 + u)
+                elif args.weighting_scheme == "cosmap":
+                    bot = (1 - 2*sigmas + 2*sigmas**2)
```
Comment (kashif) [inline examples/dreambooth/train_dreambooth_lora_sd3.py:1503]:
```suggestion
                    bot = 1 - 2 * sigmas + 2 * sigmas**2
```

File: examples/dreambooth/train_dreambooth_lora_sd3.py [hunk 2]
```diff
@@ -1483,16 +1495,15 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
...
+                    weighting = 2/(math.pi*bot)
```
Comment (kashif) [inline examples/dreambooth/train_dreambooth_lora_sd3.py:1504]:
```suggestion
                    weighting = 2 / (math.pi * bot)
```

File: examples/dreambooth/train_dreambooth_sd3.py [hunk 3]
```diff
@@ -1560,18 +1572,16 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
                 # Follow: Section 5 of https://arxiv.org/abs/2206.00364.
                 # Preconditioning of the model outputs.
                 model_pred = model_pred * (-sigmas) + noisy_model_input
-
-                # TODO (kashif, sayakpaul): weighting sceme needs to be experimented with :)
+
```
Comment (kashif) [inline examples/dreambooth/train_dreambooth_sd3.py:1575]:
```suggestion

```

File: examples/dreambooth/train_dreambooth_sd3.py [hunk 4]
```diff
@@ -1560,18 +1572,16 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
...
-                elif args.weighting_scheme == "logit_normal":
-                    # See 3.1 in the SD3 paper ($rf/lognorm(0.00,1.00)$).
-                    u = torch.normal(mean=args.logit_mean, std=args.logit_std, size=(bsz,), device=accelerator.device)
-                    weighting = torch.nn.functional.sigmoid(u)
-                elif args.weighting_scheme == "mode":
-                    # See sec 3.1 in the SD3 paper (20).
-                    u = torch.rand(size=(bsz,), device=accelerator.device)
-                    weighting = 1 - u - args.mode_scale * (torch.cos(math.pi * u / 2) ** 2 - 1 + u)
+                elif args.weighting_scheme == "cosmap":
+                    bot = (1 - 2*sigmas + 2*sigmas**2)
```
Comment (kashif) [inline examples/dreambooth/train_dreambooth_sd3.py:1581]:
```suggestion
                    bot = 1 - 2 * sigmas + 2 * sigmas**2
```

File: examples/dreambooth/train_dreambooth_sd3.py [hunk 5]
```diff
@@ -1560,18 +1572,16 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
...
+                    weighting = 2/(math.pi*bot)
```
Comment (kashif) [inline examples/dreambooth/train_dreambooth_sd3.py:1582]:
```suggestion
                    weighting = 2 / (math.pi * bot)
```

File: examples/dreambooth/train_dreambooth_lora_sd3.py [hunk 6]
```diff
@@ -1462,7 +1462,19 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
...
+                    # See 3.1 in the SD3 paper ($rf/lognorm(0.00,1.00)$).
+                    u = torch.normal(mean=args.logit_mean, std=args.logit_std, size=(bsz,), device=accelerator.device)
+                    u = torch.nn.functional.sigmoid(u)
+                elif args.weighting_scheme == "mode":
+                    u = torch.rand(size=(bsz,), device=accelerator.device)
+                    u = 1 - u - args.mode_scale * (torch.cos(math.pi * u / 2) ** 2 - 1 + u)
+                else:
+                    u = torch.rand(size=(bsz,), device=accelerator.device)
+
+
```
Comment (kashif) [inline examples/dreambooth/train_dreambooth_lora_sd3.py:1476]:
```suggestion
                    u = torch.rand(size=(bsz,), device=accelerator.device)

```

**Review by kashif — APPROVED** [review]

**Comment by chenbaiyujason** [comment]
I get the following error：
   timesteps = noise_scheduler_copy.timesteps[indices].to(device=model_input.device)
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^
RuntimeError: indices should be either on cpu or on the same device as the indexed tensor (cpu)

**Comment by sayakpaul** [comment]
@kashif I will review the changes a bit later but could you test the scripts and see if they are running without errors?

**Comment by chenbaiyujason** [comment]
This avoids errors
1476                 indices = (u * noise_scheduler_copy.config.num_train_timesteps).long()
fix:
indices = (u * noise_scheduler_copy.config.num_train_timesteps).long().to(device="cpu")

File: examples/dreambooth/train_dreambooth_lora_sd3.py [hunk 7]
```diff
@@ -1462,7 +1462,18 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
...
+                    # See 3.1 in the SD3 paper ($rf/lognorm(0.00,1.00)$).
+                    u = torch.normal(mean=args.logit_mean, std=args.logit_std, size=(bsz,), device=accelerator.device)
+                    u = torch.nn.functional.sigmoid(u)
+                elif args.weighting_scheme == "mode":
+                    u = torch.rand(size=(bsz,), device=accelerator.device)
+                    u = 1 - u - args.mode_scale * (torch.cos(math.pi * u / 2) ** 2 - 1 + u)
+                else:
+                    u = torch.rand(size=(bsz,), device=accelerator.device)
+
+                indices = (u * noise_scheduler_copy.config.num_train_timesteps).long()
```
Comment (kashif) [inline examples/dreambooth/train_dreambooth_lora_sd3.py:1476]:
```suggestion
                indices = (u * noise_scheduler_copy.config.num_train_timesteps).long().cpu()
```

File: examples/dreambooth/train_dreambooth_sd3.py [hunk 8]
```diff
@@ -1526,7 +1526,18 @@ def get_sigmas(timesteps, n_dim=4, dtype=torch.float32):
...
+                    # See 3.1 in the SD3 paper ($rf/lognorm(0.00,1.00)$).
+                    u = torch.normal(mean=args.logit_mean, std=args.logit_std, size=(bsz,), device=accelerator.device)
+                    u = torch.nn.functional.sigmoid(u)
+                elif args.weighting_scheme == "mode":
+                    u = torch.rand(size=(bsz,), device=accelerator.device)
+                    u = 1 - u - args.mode_scale * (torch.cos(math.pi * u / 2) ** 2 - 1 + u)
+                else:
+                    u = torch.rand(size=(bsz,), device=accelerator.device)
+
+                indices = (u * noise_scheduler_copy.config.num_train_timesteps).long()
```
Comment (kashif) [inline examples/dreambooth/train_dreambooth_sd3.py:1540]:
```suggestion
                indices = (u * noise_scheduler_copy.config.num_train_timesteps).long().cpu()
```

Comment (bghira) [reply examples/dreambooth/train_dreambooth_sd3.py]: `u` is on accelerator.device why

Comment (kashif) [reply examples/dreambooth/train_dreambooth_sd3.py]: cause i wanted to test and we can ask the author @Slickytail  of the PR to move it all to CPU when they respond

Comment (bghira) [reply examples/dreambooth/train_dreambooth_sd3.py]: well this particular variation of things doesn't work. the `device=accelerator.device` has to go away or be changed to assign specifically to the cpu.

Comment (bghira) [reply examples/dreambooth/train_dreambooth_sd3.py]:
```py
                    # for weighting schemes where we sample timesteps non-uniformly
                    # thanks to @Slickytail who implemented this correctly via #8528
                    if args.weighting_scheme == "logit_normal":
                        # See 3.1 in the SD3 paper ($rf/lognorm(0.00,1.00)$).
                        u = torch.normal(
                            mean=args.logit_mean,
                            std=args.logit_std,
                            size=(bsz,),
                            device="cpu",
                        )
                        u = torch.nn.functional.sigmoid(u)
                    elif args.weighting_scheme == "mode":
                        u = torch.rand(size=(bsz,), device="cpu")
                        u = (
                            1
                            - u
                            - args.mode_scale
                            * (torch.cos(math.pi * u / 2) ** 2 - 1 + u)
                        )
                    else:
                        u = torch.rand(size=(bsz,), device="cpu")

                    indices = (
                        u * noise_scheduler_copy.config.num_train_timesteps
                    ).long()
                    timesteps = noise_scheduler_copy.timesteps[indices].to(
                        device=latents.device
                    )
```

**Comment by xiaohu2015** [comment]
a question about the loss: why here we firstly convert the model prediction to x_0 to compute loss?

**Comment by kashif** [comment]
@Slickytail  yes let's keep all the `u` tensors on the CPU as the indices are on the CPU side so there is no need to set the device to gpu

**Review by sayakpaul — APPROVED** [review]
Thanks a lot for these! 

@asomoza has confirmed that these are working nicely. 

In a follow up PR, I will add your PR in the comments to honor your contributions :)

Will also make it a little utility function and move it to `training_utils.py`. 

Appreciate your help here!

**Comment by sayakpaul** [comment]
Okay I can confirm that the failing example tests are only with the latest version of `datasets`. With the 2.19.1 version, they don't appear. So, will merge the PR. Requesting @lhoestq @albertvillanova to help.

**Comment by albertvillanova** [comment]
Thanks for the ping, @sayakpaul.

In the latest `datasets` release, we had to introduce a breaking change for security reasons: https://github.com/huggingface/transformers/pull/31406
> Datasets with a Python loading script now require passing `trust_remote_code=True` to be used

**Comment by lhoestq** [comment]
I merged https://huggingface.co/datasets/hf-internal-testing/fill10/discussions/1 to fix the CI

**Pull request merged**